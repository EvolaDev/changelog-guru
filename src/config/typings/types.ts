import { ChangeLevel, ExclusionType, ServiceProvider } from './enums';

export type Changes = { [key in ChangeLevel]: string[] };
export type PluginOptionValue = string | boolean | number | string[];

export interface PluginOption {
    [key: string]: PluginOptionValue | PluginOption | PluginOption[] | undefined;
}

export interface ConfigOptions {
    provider: ServiceProvider;
    filePath: string;
    types: Map<string, ChangeLevel>;
    plugins: Map<string, PluginOption>;
    exclusions: Map<ExclusionType, string[]>;
}
