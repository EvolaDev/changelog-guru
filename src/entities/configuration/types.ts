import * as Enums from './enums';

export type Changes = { [key in Enums.Change]: string[] };
export type PluginOptionValue = string | boolean | number | string[];

export interface PluginOption {
    [key: string]: PluginOption | PluginOptionValue;
}

export interface Plugins {
    [key: string]: Plugin;
}

export interface ConfigurationOptions {
    provider: Enums.ServiceProvider;
    filePath: string;
    types: Map<string, Enums.Change>;
    plugins: Map<string, PluginOption>;
    exclusions: Map<Enums.Exclusion, string[]>;
}
