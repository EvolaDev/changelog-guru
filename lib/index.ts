import Reader from './src/reader';

const reader = new Reader();

reader.init();
reader.parse().then(() => {
    process.exit(0);
});
