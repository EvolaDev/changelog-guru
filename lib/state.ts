import * as semver from 'semver';
import { ReposListCommitsResponseItem } from '@octokit/rest';
import Section from './section';
import Process from './process';

export default class State {
    public static DEFALUT_VERSION = '1.0.0';

    private _version: string = State.DEFALUT_VERSION;
    private commits: ReposListCommitsResponseItem[] = [];
    private sections: Section[] = [];

    public set version(version: string) {
        if (!semver.valid(version)) Process.error('<package.version> is invalid (see https://semver.org/)');

        this._version = version;
    }

    public get version(): string {
        return this._version;
    }

    public addSection(name: string, patterns: string[]): void {
        this.sections.push(new Section(name, patterns));
    }

    public addCommit(commit: ReposListCommitsResponseItem): void {
        this.commits.push(commit);
    }
}
