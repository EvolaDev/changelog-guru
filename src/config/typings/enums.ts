export enum ServiceProvider {
    GitHub = 'github',
    GitLab = 'gitlab',
}

export enum ChangeLevel {
    Major = 'major',
    Minor = 'minor',
    Patch = 'patch',
}

export enum ExclusionType {
    AuthorLogin = 'authorLogin',
    CommitType = 'commitType',
    CommitScope = 'commitScope',
    CommitSubject = 'commitSubject',
}
