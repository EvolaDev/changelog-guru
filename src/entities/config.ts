import fs from 'fs';
import path from 'path';
import yaml from 'js-yaml';
import chalk from 'chalk';
import Process from '../utils/process';
import { ProviderName } from '../providers/provider';
import { ReadonlyArray, Option, OptionValue } from '../utils/types';

const $process = Process.getInstance();
const fileName = '.changelog.yaml';
const load = (p: string): ConfigOptions => yaml.safeLoad(fs.readFileSync(p, 'utf8')) || {};
const defaultConfig: ConfigOptions = load(path.join(__dirname, '../../', fileName));

export interface ConfigOptions extends Option {
    config?: string;
    types?: string[];
    plugins?: string[];
    provider?: ProviderName;
    [key: string]: Option | OptionValue;
}

export default class Config {
    public readonly plugins: ReadonlyArray<string> = defaultConfig.plugins || [];
    public readonly types: ReadonlyArray<string>;
    public readonly options: Option;
    public readonly provider: ProviderName;

    public constructor(options?: ConfigOptions) {
        $process.addTask('Config initializing');

        let config: ConfigOptions = Object.assign({}, defaultConfig, options || {});
        let { config: filePath } = config;

        if (typeof filePath === 'string') {
            filePath = path.resolve(process.cwd(), filePath);

            if (typeof filePath === 'string' && filePath.length && fs.existsSync(filePath)) {
                config = Object.assign({}, config, load(filePath));

                $process.addSubTask(`Used config file from: ${filePath}`, true);
            } else {
                $process.addSubTask(`File ${chalk.bold(fileName)} is not exists`, true);
            }
        } else {
            $process.addSubTask(`File path to ${chalk.bold(fileName)} is not specified`, true);
            $process.addSubTask(`Used default config`);
        }

        this.provider = config.provider || ProviderName.GitHub;
        this.types = config.types || [];
        this.plugins = this.plugins.concat(config.plugins || []);
        this.options = new Proxy(config as Option, {
            get(target, name, receiver): Option | undefined {
                return Reflect.get(target, name, receiver);
            }
        });

        $process.completeTask();
    }
}
