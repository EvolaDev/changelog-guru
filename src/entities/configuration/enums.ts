export enum ServiceProvider {
    GitHub = 'github',
    GitLab = 'gitlab',
}

export enum Change {
    Major = 'major',
    Minor = 'minor',
    Patch = 'patch',
}

export enum Exclusion {
    Type = 'type',
    Scope = 'scope',
    Subject = 'subject',
    Author = 'author',
}
