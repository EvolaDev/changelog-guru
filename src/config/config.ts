import { TaskTree } from 'tasktree-cli';
import { ChangeLevel, ExclusionType, ServiceProvider } from './typings/enums';
import { ConfigOptions, PluginOption } from './typings/types';
import { Provider } from '../providers/provider';
import { GitHubProvider } from '../providers/github/provider';

export class Config {
    public readonly filePath: string;
    public readonly provider: ServiceProvider;

    private types: Map<string, ChangeLevel>;
    private plugins: Map<string, PluginOption>;
    private exclusions: Map<ExclusionType, string[]>;

    public constructor(options: ConfigOptions) {
        this.provider = options.provider;
        this.filePath = options.filePath;
        this.types = options.types;
        this.plugins = options.plugins;
        this.exclusions = options.exclusions;
    }

    public async getProvider(repository: string): Promise<Provider> {
        const tasks = TaskTree.tree();
        let provider: Provider | undefined;

        switch (this.provider) {
            case ServiceProvider.GitLab:
                tasks.fail(`${ServiceProvider.GitLab} - not supported yet`);
                break;
            case ServiceProvider.GitHub:
                provider = new GitHubProvider(repository);
                break;
            default:
                tasks.fail(`Service provider not specified`);
                break;
        }

        return provider as Provider;
    }

    public getPlugin(name: string): PluginOption | undefined {
        return this.plugins.get(name);
    }

    public getPlugins(): [string, PluginOption][] {
        return [...this.plugins.entries()];
    }

    public getTypes(): [string, ChangeLevel][] {
        return [...this.types.entries()];
    }

    public getExclusions(): [ExclusionType, string[]][] {
        return [...this.exclusions.entries()];
    }
}
