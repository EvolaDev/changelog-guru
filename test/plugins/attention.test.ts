import { Task } from 'tasktree-cli/lib/task';
import { MockState } from '../__mocks__/entities/state.mock';
import { ConfigLoader } from '../../src/config/config-loader';
import AttentionPlugin from '../../src/plugins/implementations/attention/attention';
import { AttentionPluginOptions } from '../../src/plugins/implementations/attention/typings/types';
import { ChangeLevel } from '../../src/config/typings/enums';
import { DependencyRule } from '../../src/package/rules/dependency-rule';
import { DependencyRuleType } from '../../src/package/rules/typings/enums';

describe('AttentionPlugin', (): void => {
    let $loader: ConfigLoader;
    let $context: MockState;
    let $plugin: AttentionPlugin;
    let $task: Task;

    beforeEach((done): void => {
        $loader = new ConfigLoader();
        $context = new MockState();
        $plugin = new AttentionPlugin($context);
        $task = new Task('test task');

        $loader.load().then((config): void => {
            const options = config.getPlugin('attention');

            if (options) {
                $plugin.init(options as AttentionPluginOptions).then((): void => {
                    done();
                });
            } else {
                throw new Error('AttentionPlugin config not found!');
            }
        });
    });

    it('Default', (): void => {
        expect($context.getSections().length).toBe(1);
        expect($context.findSection('Important Changes')).toBeDefined();
    });

    it('Changed license', (done): void => {
        $context.setLicense('MIT', undefined);

        const section = $context.findSection('Important Changes');
        const license = $context.getLicense();

        expect(section).toBeDefined();
        expect(license).toBeDefined();

        if (section && license) {
            $plugin.modify($task).then((): void => {
                const [subsection] = section.getSections();

                expect(subsection).toBeDefined();
                expect(subsection.getName()).toBe('License');
                expect(
                    subsection
                        .getMessages()
                        .map((message): [ChangeLevel, string] => [message.getChangeLevel(), message.text])
                ).toMatchSnapshot();

                done();
            });
        }
    });

    it('Changed dependencies', (done): void => {
        $context.setPackageRule(
            new DependencyRule(
                DependencyRuleType.Dependencies,
                {
                    test: '^1.1.0',
                },
                {
                    test: '^1.0.0',
                }
            )
        );

        const section = $context.findSection('Important Changes');

        expect(section).toBeDefined();

        if (section) {
            $plugin.modify($task).then((): void => {
                const [subsection] = section.getSections();

                expect(subsection).toBeDefined();
                expect(subsection.getName()).toBe('Dependencies');
                expect(
                    subsection
                        .getMessages()
                        .map((message): [ChangeLevel, string] => [message.getChangeLevel(), message.text])
                ).toMatchSnapshot();

                done();
            });
        }
    });
});
