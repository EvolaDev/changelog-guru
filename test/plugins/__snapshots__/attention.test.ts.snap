// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AttentionPlugin Changed dependencies 1`] = `
Array [
  Array [
    3,
    "- Bumped **[test](https://www.npmjs.com/package/test/v/1.1.0)**  from \`1.0.0\` to \`1.1.0\`",
  ],
]
`;

exports[`AttentionPlugin Changed license 1`] = `
Array [
  Array [
    1,
    "Source code now under \`MIT\` license.",
  ],
]
`;
